ddtrace/debugging/_snapshot/collector.py

```diff
@@ -0,0 +1,149 @@
+
+        # TODO: Put rate limiting after condition evaluation
+        if probe.limiter.limit() is RateLimitExceeded:
+            return
+
+        snapshot = Snapshot(
+            probe=probe,
+            frame=frame,
+            thread=thread,
+            exc_info=(None, None, None),
+            context=context,
+            timestamp=time.time(),
+        )
+
+        if snapshot.evaluate(dict(args)):
+            self.snapshot = snapshot
+            self.snapshot.entry_capture = self._snapshot_encoder.capture_context(
+                args,
+                [],
+                (None, None, None),
+                level=1,  # TODO: Retrieve from probe
+            )
+
+        self.snapshot = snapshot
+
+    def exit(self, retval, exc_info):
+        # type: (Any, ExcInfoType) -> None
+        if self.snapshot is None:
+            return
+
+        self.return_value = retval
+        return self.__exit__(*exc_info)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *exc_info):
+        # type: (ExcInfoType) -> None
+        if self.snapshot is None:
+            return
+
+        # If we get here it is because we're within the rate limits and the
+        # probe condition evaluated to True.
+        args = self.args
+        _locals = (
+            [("@return", self.return_value)] if self.return_value is not NO_RETURN_VALUE and exc_info[1] is None else []
+        )  # type: List[Tuple[str, Any]]
+
+        self.snapshot.return_capture = self._snapshot_encoder.capture_context(
+            args,
+            _locals,
+            exc_info,
+            level=1,  # TODO: Retrieve from probe
+        )
+        self.collector._enqueue(self.snapshot)
+        meter.increment("encoded", tags={"probe_id": self.snapshot.probe.probe_id})
+        log.debug("Encoded %r", self.snapshot)
+
+
+class SnapshotCollector(object):
+    """Snapshot collector.
+
+    We push and do the bare minimum from the probe hook and delay the processing
+    as late as possible so to reduce interference with customer's logic.
+    """
+
+    def __init__(self, encoder):
+        # type: (BufferedEncoder) -> None
+        self._encoder = encoder
+
+    def _enqueue(self, snapshot):
+        # type: (Snapshot) -> None
+        try:
+            self._encoder.put(snapshot)
+        except BufferFull:
+            log.debug("Encoder buffer full")
+            meter.increment("encoder.buffer.full")
+
+    def push(self, probe, frame, thread, exc_info, context=None):
+        # type: (ConditionalProbe, FrameType, Thread, ExcInfoType, Optional[Context]) -> None
+        """Push hook data to the collector."""
+        snapshot = Snapshot(
+            probe=probe,
+            frame=frame,
+            thread=thread,
+            exc_info=exc_info,
+            context=context,
+            timestamp=time.time(),
+        )
+        try:
+            if snapshot.evaluate():
+                self._enqueue(snapshot)
```

Comment:
As long as the encoders themselves are using locks (which the BatchJsonEncoder does), then we should be good?)

Reply:
The snapshots will serialise all the objects referenced by `frame.f_locals`. We don't have a way to lock these data in Python so, in case there are references to objects shared by multiple threads, there is no way to prevent mutation before serialisation.

Reply:
Done